new generic reader hides error.EndOfStream, resulting in 'panic: invalid error code'

### Zig Version

0.12.0-dev.1286+cc394431a

### Steps to Reproduce and Observed Behavior

I discovered this using io.Reader().readEnum() which currently looks like this:
https://github.com/ziglang/zig/blob/a9e66ed730e17833a3f2701294c01f0b16d206e4/lib/std/io.zig#L328C8-L334C10.  You can see that it uses `any().readEnum()` where the only possible error is `InvalidValue`.

Here is a test showing what i would like to do: detect EOS.  In order to do so, I have created an alternate `readEnum()` method which doesn't use `any()`.  This test runs as expected, printing 'caught EOS'.

```zig
const std = @import("std");

const Section = enum(u8) { a, b };

fn readEnum(reader: anytype, comptime T: type) !T {
    const type_info = @typeInfo(T).Enum;
    const tag = try reader.readIntLittle(type_info.tag_type);
    return std.meta.intToEnum(T, tag);
}

test "readEnum() " {
    var fbs = std.io.fixedBufferStream("");
    const reader = fbs.reader();
    var section = readEnum(reader, Section) catch |e| switch (e) {
        error.EndOfStream => {
            std.debug.print("caught EOS\n", .{});
            return;
        },
        else => {
            std.debug.print("{s}\n", .{@errorName(e)});
            return e;
        },
    };
    _ = section;
}
```

This test attempts to do the same using GenericReader.readEnum() but fails to compile:
```console
$ zig test /tmp/tmp.zig
/tmp/tmp.zig:31:9: error: expected type 'error{InvalidValue}', found 'error{EndOfStream}'
       error.EndOfStream => {},
```

```zig
test "io.Reader().readEnum()" {
    var fbs = std.io.fixedBufferStream("");
    const reader = fbs.reader();
    var section = reader.readEnum(Section, .Little) catch |e| switch (e) {
        error.EndOfStream => {},
        else => {
            std.debug.print("{s}\n", .{@errorName(e)});
            return e;
        },
    };
    _ = section;
}
```

And if I remove `error.EndOfStream` from the switch, it compiles.  But panics when run:
```console
$ zig test /tmp/tmp.zig
Test [1/2] test.readEnum() ... caught EOS
Test [2/2] test.io.Reader().readEnum()... thread 1285307 panic: invalid error code
```



### Expected Behavior

Users of GenericReader should be able to catch all possible errors.  And it shouldn't cause panics like this when normal errors happen.

--- a/lib/std/io.zig
+++ b/lib/std/io.zig
@@ -308,7 +308,7 @@ pub fn GenericReader(
             return @errorCast(self.any().skipBytes(num_bytes, options));
         }
 
-        pub inline fn isBytes(self: Self, slice: []const u8) Error!bool {
+        pub inline fn isBytes(self: Self, slice: []const u8) NoEofError!bool {
             return @errorCast(self.any().isBytes(slice));
         }
 
@@ -320,7 +320,7 @@ pub fn GenericReader(
             return @errorCast(self.any().readStructBig(T));
         }
 
-        pub const ReadEnumError = Error || error{
+        pub const ReadEnumError = NoEofError || error{
             /// An integer was read, but it did not match any of the tags in the supplied enum.
             InvalidValue,
         };

--- a/lib/std/io/test.zig
+++ b/lib/std/io/test.zig
@@ -182,3 +182,16 @@ test "updateTimes" {
     try expect(stat_new.atime < stat_old.atime);
     try expect(stat_new.mtime < stat_old.mtime);
 }
+
+test "GenericReader methods can return error.EndOfStream" {
+    // https://github.com/ziglang/zig/issues/17733
+    var fbs = std.io.fixedBufferStream("");
+    try std.testing.expectError(
+        error.EndOfStream,
+        fbs.reader().readEnum(enum(u8) { a, b }, .Little),
+    );
+    try std.testing.expectError(
+        error.EndOfStream,
+        fbs.reader().isBytes("foo"),
+    );
+}

